### ISSUE TITLE
Custom CodeWorld configurations

### ISSUE THREAD

**Issue by cdsmith** [google/codeworld#138]
A few people have expressed interest in more CodeWorld configurations.  There are currently two: the core configuration, and the Haskell configuration.

Ideas for more configurations:
- Include a library for compass-and-ruler constructions a la geogebra created by Fernando Alegre.
- Include an embedded DSL for physics based on https://hackage.haskell.org/package/learn-physics by @walck
- Fixed long-term stable API versions for codeworld-base or codeworld-api, which would be guaranteed to remain fixed for a class, or even for a multi-year sequence.

Configurations would need:

1. Libraries that should be installed and made available.
2. Customized compiler options.
3. Custom error message processing.  (And if we start preprocessing source code, custom plugins for that, too.)

There are open questions about the UI:
* How do you create a configuration?  By sending a pull request and asking for a new site release?
* How do you select a configuration to use for your project?  Currently, each configuration gets a different base URL and hash prefix.

**Comment by alphalambda** [comment]
I think this issue is actually several separate issues.
We may need to make a distinction between regular Haskell packages using the standard prelude and standard curried style on one hand, and modules that depend on the CodeWorld prelude and are written in f(x,y,z) style on the other hand. Let me refer to the former as Haskell modules and to the latter as CodeWorld modules. So, configuration could mean:

1. Being able to import custom Haskell modules pre-installed in the system
2. Being able to import custom CodeWorld modules pre-installed in the system
3. Being able to import CodeWorld modules exported by other (or same) users

Modules of type 1 and 2 could be handled similarly, except for the fact that modules of type 1 would typically reside on Hackage or similar and be installed with cabal or stack, while it is not clear to me whether uploading modules of type 2 to Hackage makes so much sense, as probably a CodeWorld-specific repository would make more sense.

**Comment by cdsmith** [comment]
Building libraries within CodeWorld (your type 3) is covered by issue #12.  Building that right (without making the platform harder to use for new students) seems much harder.

Narrowing the scope to pre-installed packages, I'm planning on solving this by generalizing the approach already taken to differentiate between http://code.world and http://code.world/haskell -- these URLs use the same basic site logic, but different project stores, libraries, compiler options, and error processing.  The trick is that this distinction is currently made in if statements throughout the code, so that needs cleaning up so that a new configuration can be added in a single place.

Assuming all libraries are built and installed in the GHCJS package database as part of the site build, then whether the module is installed from Hackage or something more local to CodeWorld (for now, that means residing locally in the CodeWorld project), isn't really a problem.  Just a detail of the build script.

**Comment by alphalambda** [comment]
I'm not sure I can made a fleshed out proposal yet, because I need more thought about it, but I I added some comments to the issues raised in #783 to get the discussion going.

- _Avoiding too much additional complexity in the UI. It's important to avoid losing too much class time to students setting up the right configuration for their projects._

I agree, but a little configuration should not be too distracting. I'm not sure whether anything is needed beyond the current mode=whatever in the URL.

- _Coming up with a communication story. It's already confusing that "CodeWorld" is a name for both the environment, and the simplified language variant. We probably would end up devising new names for the different supported language configurations._

I would distinguish between the CodeWorld System (mostly account,auth,compiler,server,and partially api) and the CodeWorld Library (base and partially api). I'm not sure how to classify funblocks or game-* wrt this issue. Also, I don't know how you feel about renaming the CodeWorld Library components into something like Middle School Prelude. That way, we could talk about what belongs only in the Middle School Prelude, what belongs only in a High School Prelude, and what could be in a common library.

- _Abstracting over all the differences. There are many places where haskell-mode and codeworld-mode differ: compiler flags, libraries, editor configuration, error message handling, debugger logic, etc. We'd want some kind of module surface for a configuration to handle all of these._

Ideally, all libraries could be available in all modes, even if some renaming must take place, as their availability does not limit any use in any way.  I sometimes wished I could use the codeworld-base Prelude in haskell mode, so that I could reuse some code I wrote.

As for the rest, I think it is easy. Many compiler flags are about {-# language pragmas, so that is not a big deal as long as pragmas can still be used. As for editor configuration and extra error handling, there are probably only 2 possible behaviors: either work like the codeworld mode or work like the haskell mode.
So a simple flag could be used to select between those.

- _Namespacing and ownership of configurations. Should configurations be owned by the CodeWorld project, which implies a high barrier to entry for adding a new one? Or should they be owned by non-privileged users, which implies some kind of name registry, or unique ids, or something, and more complex processes for teachers to share them with students?_

The long-term solution may require some kind of registry, but for now, maybe just having one or two extra modes is all we need.

In particular, a High School Prelude that places the trade-off between simplicity and features at a different point from the Middle School Prelude would solve most problems I currently have. For example, there
are features that I actually want to teach to high schoolers, such as function composition and the `(.)` operator, which may not be necessary in the Middle School Prelude, and so I understand why they are being deprecated in the Middle School Prelude, but I also would like to keep them in a High School Prelude.

**Comment by cdsmith** [comment]
Okay, this sounds like a good start.  I don't want to completely ignore the possibility of teacher-defined configurations, which have a lot of potential.  But sure, a good place to start might be a small collection of baked-in modes.

Off the top of my head, I'm not a big fan of "middle school prelude" and "high school prelude", because different populations will have different needs that vary based on more than just age.  An elective after-school middle school activity in a private school in Silicon Valley might want to cover topics considerably more advanced than a math teacher who is devoting one day a week of a core math class in a high-poverty rural public school, even if the students at the former are the same age.  But that need not block more thinking.  Naming doesn't have to come first.

I don't understand what you mean by using the codeworld-base prelude in haskell-mode.  The main difference between mode=codeworld and mode=haskell is just which library is exposed.  The conventions around built-in types and currying are different enough that no single library is likely to be usable by kids in both environments.

There's definitely room for more behaviors on error sanitation (adding new patterns that are specific to the types and such for a given mode) and editor configuration, and the hope is to avoid kids having to start their code with a long list of language pragmas.  But okay, I agree we can postpone these for later in the process; if the modes are hard-coded, then it's easy enough to define more lists and choose between them in the compiler wrapper.

**Comment by alphalambda** [comment]
I agree with all your points, except for the utility of using codeworld-base in haskell-mode. Even in haskell-mode, I found it useful to use just Number instead of Int and Double, to have Text by default instead of String, to use # for 1-based indices and to be able to reuse code I made for codeworld-mode without having to make lots of nitty-gritty currying changes just because the calling convention is different. This is also a hindrance for advanced students who want to transition to haskell-mode but still be able to reuse the codes they did in codeworld-mode and build on them. Having their old code work without changes in haskell-mode helps them make the transition more smoothly.

I did some local tests, and it worked, after renaming some modules, so there is no technical reason why this should not be available. As for whether it should be allowed, I would prefer the platform itself to be agnostic, so that different teachers or curriculum writers can make different choices.

I agree that the other way, using haskell-mode libraries in codeworld-mode, seems less useful, because it narrows instead of widening the options.

Should this point be submitted as a separate issue?
